Add Async.await() for non-blocking condition waiting

mfateev opened the pull request:
## Summary
- Add `Async.await(Supplier<Boolean>)` and `Async.await(Duration, Supplier<Boolean>)` methods for non-blocking condition waiting that return a `Promise`
- Add comprehensive tests for cancellation behavior and condition exception handling
- Add cross-reference `@see` Javadoc tags between `Workflow.sleep`, `Workflow.newTimer`, `Workflow.await`, and `Async.await` to help developers discover blocking vs non-blocking alternatives

## Test plan
- [x] Unit tests for `Async.await()` with immediate condition satisfaction
- [x] Unit tests for timed `Async.await()` with timeout expiration
- [x] Unit tests for cancellation behavior (CanceledFailure propagation)
- [x] Unit tests for condition throwing exceptions
- [x] All existing AsyncAwaitTest and WorkflowTest tests pass

mfateev committed:
Add Async.await() for non-blocking condition waiting

Introduces Async.await() methods that return a Promise instead of blocking,
allowing workflows to wait for conditions asynchronously. This enables
concurrent condition waiting and Promise composition with anyOf/allOf.

Key changes:
- Add Async.await(Supplier<Boolean>) returning Promise<Void>
- Add Async.await(Duration, Supplier<Boolean>) returning Promise<Boolean>
- Implement condition watchers in SyncWorkflowContext evaluated via
  beforeThreadsWakeUp callback in DeterministicRunner
- Support cancellation through CancellationScope
- Add comprehensive tests including chaining, anyOf/allOf, and cancellation

mfateev committed:
Add tests for Async.await() timed cancellation and condition exceptions

- Add testTimedAsyncAwaitCancellation: verifies cancellation of the timed
  Async.await(Duration, Supplier) variant via CancellationScope
- Add testAsyncAwaitConditionThrows: verifies that exceptions thrown by
  conditions during subsequent evaluations complete the promise exceptionally

mfateev committed:
Add cross-reference @see tags between related wait/sleep methods

Link Workflow.sleep, Workflow.newTimer, Workflow.await, and Async.await
in Javadoc to help developers discover blocking vs non-blocking alternatives.

mfateev committed:
Add AwaitOptions with timerSummary for Async.await() and Workflow.await()

Add overloads for await methods that accept AwaitOptions, allowing users
to specify a timer summary that appears in the workflow history UI/CLI.

- Add AwaitOptions class with timerSummary field
- Add Async.await(Duration, AwaitOptions, Supplier) for non-blocking waits
- Add Workflow.await(Duration, AwaitOptions, Supplier) for blocking waits
- Add test validating timer summary is set in workflow history

mfateev commented on temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptor.java:
```diff
@@ -742,6 +742,26 @@ public DynamicUpdateHandler getHandler() {
+   *     expired, or exceptionally with CanceledFailure if the enclosing CancellationScope is
+   *     canceled
+   */
+  Promise<Boolean> awaitAsync(Duration timeout, Supplier<Boolean> unblockCondition);
```
Added timerSummary to await

mfateev committed:
Simplify awaitAsync and condition watcher implementation

- Remove redundant condition checks from awaitAsync (registerConditionWatcher
  already evaluates inline)
- Remove AtomicBoolean from non-timed awaitAsync (cancellation handled by
  thenApply callback)
- Simplify registerConditionWatcher to return Runnable cancel handle
- Remove ConditionWait class (no longer needed)
- Simplify ConditionWatcher to only store condition (promise completion
  handled in wrapped condition)
- Timed awaitAsync delegates to version with AwaitOptions
- Update documentation to clarify inline evaluation behavior

maciejdudko reviewed (changes requested):
The existing `await` overloads always requires reason. `awaitAsync` should match that.

I don't think it's worth introducing `AwaitOptions`. `await` already takes the reason as direct argument, and it's more ergonomic this way. If there's ever a need for additional options (unlikely), we can add the `AwaitOptions` overloads then. For now, `awaitAsync` should have `String reason` as a direct argument too.

In short - there should be just these two overloads:
```
public static Promise<Void> awaitAsync(String reason, Supplier<Boolean> unblockCondition)
public static Promise<Boolean> awaitAsync(Duration timeout, String reason, Supplier<Boolean> unblockCondition)
```

I'm still reviewing the rest of the code, but wanted to highlight this comment early.

mfateev commented:
Will remove async options. Hold reviewing the code. I want to change it to execute the condition in a workflow thread for consistency with other callbacks.

mfateev committed:
Replace AwaitOptions with String reason parameter

Per reviewer feedback, remove AwaitOptions class and use String reason
as a direct argument for awaitAsync methods. The reason parameter is
used for debugging, stack traces, and timer summary.

API changes:
- awaitAsync(String reason, Supplier<Boolean>)
- awaitAsync(Duration timeout, String reason, Supplier<Boolean>)

Implementation uses RepeatableWorkflowThread that:
- Evaluates conditions repeatedly until they return true
- Reports progress only when condition is satisfied
- Conditions must be read-only (observe state, not modify it)

mfateev committed:
Use timerSummary parameter only for timed await

- Remove reason parameter from awaitAsync without timeout
- Rename reason to timerSummary for timed await methods
- timerSummary is used for the timer created by await with timeout

API:
- Async.await(Supplier<Boolean>) - no timer
- Async.await(Duration, Supplier<Boolean>) - timer, no summary
- Async.await(Duration, String timerSummary, Supplier<Boolean>) - timer with summary

mfateev committed:
Add test for cancellation while condition is blocked

Test verifies that cancellation properly propagates when the condition
function inside Async.await() is blocked on an internal Workflow.await().

mfateev committed:
Remove trivial comments, fix flaky test

- Remove comments that trivially explain the code that follows
- Keep only comments that explain non-obvious design decisions
- Fix testMultipleConcurrentRepeatableThreads: use >= assertions since
  event loop may run extra iterations when threads complete

mfateev committed:
Fix getCancellationRequest() to cache promise

Cache the cancellation promise in RepeatableWorkflowThread following
the same pattern used in CancellationScopeImpl. This ensures callers
get the same promise instance and it gets completed when cancel() is called.

mfateev committed:
Add .serena to .gitignore

mfateev committed:
Remove .serena from version control

mfateev committed:
Remove beforeThreadsWakeUp callback and condition watcher mechanism

With RepeatableWorkflowThread, condition evaluation now happens inside
workflow threads themselves, making the beforeThreadsWakeUp callback
and condition watcher mechanism obsolete.

- Remove beforeThreadsWakeUp parameter from DeterministicRunner
- Remove conditionWatchers, registerConditionWatcher, evaluateConditionWatchers
  from SyncWorkflowContext
- Remove ConditionWatcher class
- Remove getBeforeThreadsWakeUpCallback method
- Remove related tests

mfateev committed:
Clean up imports in RepeatableWorkflowThread

Use imported class names instead of fully qualified names for
CompletablePromise, Promise, and Workflow classes.

mfateev committed:
Remove unused imports

mjameswh commented on temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunner.java:
```diff
@@ -92,6 +93,24 @@ static DeterministicRunner newRunner(
+   * runUntilBlocked() call. The thread completes when the condition returns true or when
+   * cancelled/destroyed.
+   *
+   * <p>This is used for async await where the condition can contain blocking operations like
```
Here, the Javadoc indicates that the condition provided to `newRepeatableThread` can contain blocking operations. But the same method in DeterministicRunnerImpl says the opposite.

Can you confirm which one is in line your your now-currrent design?
